### PR TITLE
Contact list

### PR DESCRIPTION
**Pull request by TdVBd**
Implemented the utilitary database functions for the Contact List feature aswell as the corresponding data class and viewmodel
Modified the DirectMessageScreen, now when starting a new chat with a user from the userlist, a new contact is created between it and the current user

### CODE REVIEW

File: app/src/main/java/com/github/se/studybuddies/data/Contact.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+data class Contact(val id: String, val members: Pair<String, String>, val showOnMap: Boolean) {
+  companion object {
+    fun empty(): Contact {
+      return Contact(id = "", members = Pair("", ""), false)
+    }
+  }
+}
+
+class ContactList(private val contacts: List<Contact>) {
+  fun getAllTasks(): List<Contact> {
```
Comment (hanelmouta) [inline app/src/main/java/com/github/se/studybuddies/data/Contact.kt:12]: getAllTasks should be renamed to getAllContacts for clarity.

Comment (hanelmouta) [inline app/src/main/java/com/github/se/studybuddies/data/Contact.kt:12]: why do you have this companion function ?

File: app/src/main/java/com/github/se/studybuddies/database/DatabaseConnection.kt [hunk 2]
```diff
@@ -1112,6 +1125,101 @@ class DatabaseConnection {
...
+      val members = document.get("members") as? Pair<String, String> ?: Pair("", "")
+      val showOnMap = document.get("showOnMap") as Boolean
+      Contact(contactUID, members, showOnMap)
+    } else {
+      Log.d("MyPrint", "contact document not found for contact id $contactUID")
+      Contact.empty()
+    }
+  }
+
+  suspend fun createContact(otherUID: String) {
```
Comment (hanelmouta) [inline app/src/main/java/com/github/se/studybuddies/database/DatabaseConnection.kt:1166]: I think you can create two time the same contact is there a way to ensure that only one contact of the same person is created ?

File: app/src/main/java/com/github/se/studybuddies/viewModels/ContactViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.github.se.studybuddies.data.Contact
+import com.github.se.studybuddies.database.DatabaseConnection
+import kotlinx.coroutines.launch
+
+class ContactViewModel(private val contactID: String? = null) : ViewModel() {
```
Comment (hanelmouta) [inline app/src/main/java/com/github/se/studybuddies/viewModels/ContactViewModel.kt:11]: Can't you regroup the two viewModels?

**Review by hanelmouta — APPROVED** [review]
Hi , I reviewed your pr , you did a really good job. Your code is clean and well written . They are no major change in the ui but I check that everything is visible in the firebase and it is.  There is a small issues and it is that you can create a contact twice of the same person.

File: app/src/main/java/com/github/se/studybuddies/data/Contact.kt (hunk 1 above)
Comment (hanelmouta) [inline app/src/main/java/com/github/se/studybuddies/data/Contact.kt:12]: The empty() method in the companion object of Contact might be redundant if not used.

Comment (TdVBd) [reply app/src/main/java/com/github/se/studybuddies/data/Contact.kt:12]: you can check its usages

File: app/src/main/java/com/github/se/studybuddies/database/DatabaseConnection.kt (hunk 2 above)
Comment (TdVBd) [reply app/src/main/java/com/github/se/studybuddies/database/DatabaseConnection.kt:1166]: this bug has now been solved

File: app/src/main/java/com/github/se/studybuddies/viewModels/ContactViewModel.kt (hunk 3 above)
Comment (TdVBd) [reply app/src/main/java/com/github/se/studybuddies/viewModels/ContactViewModel.kt:11]: yes i might do that

**Pull request merged**